apps/test/unit/code-studio/pd/application/TeacherApplicationIntegrationTest.js

```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+import {expect} from 'chai';
+import {mount} from 'enzyme';
+import sinon from 'sinon';
+import {PageLabels} from '@cdo/apps/generated/pd/teacherApplicationConstants';
+import TeacherApplication from '@cdo/apps/code-studio/pd/application/teacher/TeacherApplication';
+import firehoseClient from '@cdo/apps/lib/util/firehose';
+
+describe('TeacherApplication', () => {
+  const fakeOptionKeys = Object.values(PageLabels).reduce(
+    (acc, cur) => acc.concat(Object.keys(cur)),
+    []
+  );
+  const fakeOptions = fakeOptionKeys.reduce(
+    (acc, cur) => ({...acc, [cur]: ['1', '2', '3']}),
+    {}
+  );
+  const defaultProps = {
+    apiEndpoint: '/path/to/endpoint',
+    options: fakeOptions,
+    accountEmail: 'user@email.com',
+    userId: 1
+  };
+
+  beforeEach(() => {
+    sinon.stub(firehoseClient, 'putRecord');
+    sinon
+      .stub(window.sessionStorage, 'getItem')
+      .withArgs('TeacherApplication')
+      .returns(JSON.stringify({}));
+    sinon.stub(window.sessionStorage, 'setItem');
+    window.ga = sinon.fake();
+  });
+
+  afterEach(() => {
+    firehoseClient.putRecord.restore();
+    window.sessionStorage.getItem.restore();
+    window.sessionStorage.setItem.restore();
+    window.ga = undefined;
+  });
+
+  it('Sends firehose event on initialization and save', () => {
+    // Also calls firehose event on submit, but we reload the page on submit
+    // The page reload can't be stubbed because `window.location.reload`
+    // is not writable nor configurable: Object.getOwnPropertyDescriptor(window.location, 'toString')
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty#modifying_a_property
```

Comment:
I spent too long on this 😅 . We could do the window.location.reload on a separate function that we stub out, but then we'd have to export that function for the test here. I don't think there's a "good" way to test this––open to ideas

Reply:
so we actually do have a util already specifically for this purpose: https://github.com/code-dot-org/code-dot-org/blob/1274ce4088eb23433ba676c0871eb7950fd60b61/apps/src/utils.js#L725-L727 that is what I would recommend if you want to verify that a page reload has happened in a unit test.

Reply:
Ooo excellent––will update this!